### PR TITLE
Deprecate `assert_template` and `assigns()`.

### PR DESCRIPTION
**Pull request by tgxworld**
Related: https://github.com/rails/rails/issues/18950.

cc/ @dhh

### CODE REVIEW

File: actionpack/test/controller/action_pack_assertions_test.rb [hunk 1]
```diff
@@ -304,14 +304,6 @@ def session_does_not_exist
     assert session.empty?
   end
 
-  def test_render_template_action
-    process :nothing
-    assert_template nil
-
-    process :hello_world
-    assert_template 'hello_world'
-  end
```
Comment (tgxworld) [inline actionpack/test/controller/action_pack_assertions_test.rb]: This is a duplicated test.

File: actionpack/test/controller/integration_test.rb [hunk 2]
```diff
@@ -315,8 +315,6 @@ def test_opens_new_session
     session1 = @test.open_session { |sess| }
     session2 = @test.open_session # implicit session
 
-    assert_respond_to session1, :assert_template, "open_session makes assert_template available"
-    assert_respond_to session2, :assert_template, "open_session makes assert_template available"
```
Comment (tgxworld) [inline actionpack/test/controller/integration_test.rb]: Duplicated in https://github.com/tgxworld/rails/blob/deprecated_assert_template/actionpack/test/dispatch/template_assertions_test.rb#L66

Comment (dhh) [inline actionpack/test/controller/integration_test.rb]: Would be nice to have both assert_template and assigns() available in a
plugin, so people who like that kind of thing can continue using it.

Comment (tgxworld) [inline actionpack/test/controller/integration_test.rb]: Publish the plugin as a gem?

Date: Wed, 13 May 2015 07:14:31 -0700

Comment (dhh) [inline actionpack/test/controller/integration_test.rb]: Yes, a separate gem that we can point the deprecations to.

File: guides/source/testing.md [hunk 3]
```diff
@@ -580,11 +579,11 @@ To test AJAX requests, you can specify the `xhr: true` option to `get`, `post`,
 `patch`, `put`, and `delete` methods:
 
 ```ruby
-test "ajax request responds with no layout" do
```
Comment (simi) [inline guides/source/testing.md]: I this name is still valid for this test.

Comment (tgxworld) [inline guides/source/testing.md]: It doesn't render a layout but I think we shouldn't be encouraging others to test for it.

File: actionpack/test/controller/integration_test.rb (hunk 2 above)
Comment (tgxworld) [inline actionpack/test/controller/integration_test.rb]: Sure I'll look into it.

**Comment by tgxworld** [comment]
@dhh @matthewd I've deprecated both `assigns()` and `assert_template` and extraced both methods into a plugin. I'll still have to publish the gem though.

**Comment by tgxworld** [comment]
Done! Gem published at https://rubygems.org/gems/rails-controller-testing.

File: actionpack/lib/action_controller/template_assertions.rb [hunk 4]
```diff
@@ -102,6 +102,11 @@ def reset_template_assertion
     #   # assert that the "_customer" partial was rendered with a specific object
     #   assert_template partial: '_customer', locals: { customer: @customer }
     def assert_template(options = {}, message = nil)
+      ActiveSupport::Deprecation.warn(
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: We should raise a NoMethodError and point to the gem so we can remove all the implementation

File: actionpack/lib/action_dispatch/testing/test_process.rb [hunk 5]
```diff
@@ -5,6 +5,12 @@
 module ActionDispatch
   module TestProcess
     def assigns(key = nil)
+      ActiveSupport::Deprecation.warn(
```
Comment (rafaelfranca) [inline actionpack/lib/action_dispatch/testing/test_process.rb]: same about NoMethodError

File: actionpack/lib/action_controller/template_assertions.rb (hunk 4 above)
Comment (tgxworld) [inline actionpack/lib/action_controller/template_assertions.rb]: Won't we need to deprecate it first?

Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: If we are extracting a gem there is no need. The `NoMethodError` message should tell about the gem and people can get it work adding the gem to their Gemfile.

Comment (tgxworld) [inline actionpack/lib/action_controller/template_assertions.rb]: Do you want me to transfer the Gem to Rails? Although Rails isn't supporting these two methods, it seems weird that we're pointing to my personal repository.

Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: :+1: transfer to Rails so I can give you commit access.

Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: And also we make it official

Comment (tgxworld) [inline actionpack/lib/action_controller/template_assertions.rb]: Hmm I can't transfer without admin rights. Better to fork in this case?

Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: transfer to me so I can transfer to Rails.

Comment (tgxworld) [inline actionpack/lib/action_controller/template_assertions.rb]: Thanks! I'll remove the method soon.

Comment (rafaelfranca) [inline actionpack/lib/action_controller/template_assertions.rb]: Don't remove the methods, just change their implementation to:

```ruby
def assert_template(*)
  raise NoMethodError, "assert_template was extracted to a gem. To continue using it use the ..."
end
```

Comment (tgxworld) [inline actionpack/lib/action_controller/template_assertions.rb]: Ok I'll remove the internal tests in this case.

**Comment by tgxworld** [comment]
Updated but there is some problem on Travis that I isn't related.

**Comment by rafaelfranca** [comment]
Great work! Could you add a CHANGELOG entry?

**Comment by tgxworld** [comment]
> Great work! Could you add a CHANGELOG entry?

O yeah forgot about it XD. Done :smile:

File: actionpack/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,9 @@
+*   Deprecate `assigns` and `assert_template`.
```
Comment (rafaelfranca) [inline actionpack/CHANGELOG.md]: I think it is better to tell that these methods were extracted to the gem and point to it on the changelog.

File: actionpack/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,10 @@
+*   Remove `assigns` and `assert_template`. Both methods have been extracted
+    into a gem at https://github.com/rails/rails-controller-testing.
```
Comment (tgxworld) [inline actionpack/CHANGELOG.md]: @rafaelfranca Is this ok? :smile:

**Comment by kirs** [comment]
:+1: great job!

**Pull request merged**